contrib/babelfishpg_common/src/datetime.c

```diff
@@ -679,6 +682,596 @@ datetime_pl_datetime(PG_FUNCTION_ARGS)
+							if (regexec(&time_regex, field[i], 0, NULL, 0) != 0)
+								TIME_IN_ERROR();
+							regfree(&time_regex);
+						}
+						else
+						{
+							/*
+							 * If text month is present then supported date formats are
+							 * "DD-MON-YYYY", "YYYY-MON-DD".
+							 * Note: The year can be of 1/2/4 digits and DD can be of 1/2 digits.
+							 */
+							pattern = "^([0-9]{4}|[0-9]{1,2})[-/.]([a-z]*)[-/.]([0-9]{4}|[0-9]{1,2})$";
+							if (regcomp(&time_regex, pattern, 1) != 0)
+								ereport(ERROR,
+											(errcode(ERRCODE_INTERNAL_ERROR),
+											errmsg("time format internal error")));
+							if (regexec(&time_regex, field[i], 0, NULL, 0) != 0)
+								TIME_IN_ERROR();
+							regfree(&time_regex);
+						}
+					}
+					break;
+
+				case DTK_TIME:
+					/*
+					 * If the modified_str is of format "<DATE>T<TIME>", then the
+					 * <TIME> field should be of format hh:mm:ss[.sss]
+					 * where hh should definitely be 2 digit.
+					 *
+					 * For example if the field is "11:60:00" then an error will be thrown,
+					 * and no error will be thrown if field is "11:59:59".
+					 */
+					if (i-1 == 1 && field[1] && pg_strcasecmp(field[1], "t") == 0)
+					{
+						pattern = "^([0-1][0-9]|2[0-3])(:[0-5]?[0-9]:([0-5]?[0-9]|[0-5]?[0-9].[0-9]{1,9}))?$";
+						if (regcomp(&time_regex, pattern, REG_EXTENDED) != 0)
+							ereport(ERROR,
+										(errcode(ERRCODE_INTERNAL_ERROR),
+										errmsg("time format internal error")));
+						if (regexec(&time_regex, field[i], 0, NULL, 0) != 0)
+							TIME_IN_ERROR();
+						regfree(&time_regex);
+					}
+					/*
+					 * For all other time fields check if <TIME> is in hh:mm[:ss[[:.]nnn]] format.
+					 * For example if field is "2:3" then no error will be thrown,
+					 * and if the field is "2:3.4" then an error will be thrown because when
+					 * there is an microseconds mentioned then the hours field should also need to be mentioned.
+					 */
+					else
+					{
+						pattern = "^([0-1]?[0-9]|2[0-3]|[0-9])(:[0-5]?[0-9])(:[0-5]?[0-9]|:[0-5]?[0-9][:.][0-9]{1,9})?$";
+						if (regcomp(&time_regex, pattern, REG_EXTENDED) != 0)
+							ereport(ERROR,
+										(errcode(ERRCODE_INTERNAL_ERROR),
+										errmsg("time format internal error")));
+						if (regexec(&time_regex, field[i], 0, NULL, 0) != 0)
+							TIME_IN_ERROR();
+						regfree(&time_regex);
+						res = makeStringInfo();
+						/*
+						 * For inputs of format "HH:MM:SS:nnn" will be converted to "HH:MM:SS.nnn".
+						 * where nnn ranges from 0 to 999.
+						 */
+						for(j = 0; j <len; j++)
+						{
+							if(field[i][j]== ':')
+							{
+								time_count++;
+								if(time_count == 3)
+								{
+									field[i][j]='.';
+								}
+							}
+							if(time_count == 3 && field[i][j]!= '.')
+							{
+								appendStringInfo(res, "%c", field[i][j]);
+							}
+						}
+						temp_time = res->data;
+						pfree(res);
+						if(time_count >2)
+						{
+							temp_field = pstrdup(field[i]);
```

Comment:
This can be resolved in 2 ways:
1. Don't allocate to temp_field and directly use the field value while appending to res.
2. pfree the allocate memory and assign new value.
I prefer choosing the 1st way as this field value is not being used anywhere in that case.